### ISSUE TITLE
Leela Zero not good at ladder?

### ISSUE THREAD

**Issue by trunterzx** [leela-zero/leela-zero#1482]
Was trying out some handicap games with Leela Zero, and I was surprised that Leela Zero ran at B when I atari at A. I thought that it would do a ladder breaker after running 1 stone, but it actually ran all the way until the end!!

Is this a bug, or is Leela Zero bad with handicap games?

<img width="1155" alt="screen shot 2018-05-26 at 6 36 44 pm" src="https://user-images.githubusercontent.com/36939338/40575252-1c375f84-6114-11e8-8508-4476da6102e0.png">

**Comment by jokkebk** [comment]
Depending on the amount of visits (how many were you on?), ladders are a small problem (large amount of visits/time) or a big problem (less than 5000 visits definitely). The issue is more apparent in handicap games where there are "ladder breakers" for black in place. This is not a Leela-specific issue, I recall FB ELF network which is a lot stronger, is relatively weaker in reading ladders.

With high visit counts (say 20-30k) it starts being less of a problem. Also, the network is not trained for handicap, it just happens to be quite good at it as well. 😄

**Comment by tterava** [comment]
Leela Zero outputs two values which are used to determine which move should be played:
1. All the moves LZ thinks are important to explore. Most of the time these are what LZ thinks are the best moves to play, but often they are moves which are used to determine if a ladder works, if a group can be killed or making sure some tesuji doesn't work.
2. Evaluation of the given board position.

In handicap games no matter what board position you give, LZ tends to give a very low win percetange (less than 1%). This is correct, because LZ would pretty much never win against itself if one side gets sizable handicap. This means that the second part of the network becomes completely useless: If all positions are losing then it doesn't matter what what you play. What then happens is that LZ starts to play only moves that it thinks are worth considering. Ladders are very often important to read out, so they get high prior value. In normal situations LZ's evaluation would stop it from playing into bad ladder but when the evaluation is 0% regardless, it ends up playing it almost inevitably.

TL;DR LZ is not very good at handicap as it wasn't designed for that.

**Comment by trunterzx** [comment]
Thanks for the explanation!

However, I still think it has something to do with this particular shape that causes this issue. I've watched 2 games played by ELF Go that ended up in the same corner joseki, and ELF also ran the ladder. The difference was that ELF ran 1 or 2 moves before stopping to play elsewhere.

Makes me wonder whether running the ladder is really a good move, or there is something special about this joseki that causes bugs...

**Comment by tterava** [comment]
It's not a bug. Reading ladders requires a decent amount of playouts (500 - 1000 minimum). In some situations it might need more if there are other interesting moves that need to be considered as well.

Try this:
1. Make it an even game.
2. Give LZ 5000 playouts per move or more.
3. Make sure the evaluation is between 20% and 80%.
4. Don't mess with the fuseki LZ wants to play.

You will find out that LZ is extremely unlikely to play into a ladder. Just remember that LZ is a tool tailored for one job and one job only: Winning an even game of Go when it is allowed to play all stones for one side. If you manually put LZ into situations where it would never be if it had played from the start, it's not guaranteed to perform well. It might make assumptions that are not true anymore.

**Comment by trunterzx** [comment]
I see. Well if it's not a bug then it seems like a weakness of LZ worth exploring. Humans can force LZ(or other AIs) into a ladder that is difficult to calculate, and if the number of playout is not enough LZ has a high chance of running the ladder, which leads to an advantage for the human.

**Comment by tterava** [comment]
LZ does explore it on its own. All self play and match games are played with 3200 visits, so that's the minimum which should be used in normal games as well.

The appropriate way for LZ to deal with ladders is to give the ladder moves high prior if the ladder is important, then let the playouts flow into that move so we know whether the ladder works or not. LZ already knows how to do this. In fact it has known how to do this for a long time now. It only has problems with people put it into situations where it would never be in normal games or don't give it enough playouts to read the ladders.

**Comment by betafj** [comment]
"I see. Well if it's not a bug then it seems like a weakness of LZ worth exploring. Humans can force LZ(or other AIs) into a ladder that is difficult to calculate, and if the number of playout is not enough LZ has a high chance of running the ladder, which leads to an advantage for the human." 

In even games it is very very difficult to force LZ into a ladder that is difficult to calculate. I don't think any human player is capable of doing this if the latest LZ network has a sufficient number (see previous posts for an exact number) of visits.

**Comment by gcp** [comment]
> Humans can force LZ(or other AIs) into a ladder that is difficult to calculate, and 

It is very tricky to force LZ into a situation with ladders that are difficult to calculate. Haylee has commented on this on some of her videos.

>if the number of playout is not enough LZ

If you run the engine at a low level, it plays weaker. This is not a surprise.

**Comment by Nazgand** [comment]
Perhaps this could be fixed by collecting board positions in which LeelaZero chooses to play out a bad ladder, and use them as starting points in AutoGTP. This may not help much if the ladder is not played out under self-play.
LeelaZero is indeed suboptimal in handicap games or any game in which it believes it is unlikely to win. A `score head' proposal may fix this. https://github.com/gcp/leela-zero/issues/1470#issuecomment-392452882

**Comment by sethtroisi** [comment]
ELF mentioned similar problems and constructed such a set of positions see there recent paper.

closing older issues, no discussion for >6 months
